app/modules/sntppkt.c

```diff
@@ -0,0 +1,421 @@
+#include <stdlib.h>
+#include "user_modules.h"
+#include "lwip/dns.h"
+#include "task/task.h"
+#include "user_interface.h"
+
+#define max(a,b) ((a < b) ? b : a)
+
+#define NTP_PORT 123
+#define NTP_ANYCAST_ADDR(dst)  IP4_ADDR(dst, 224, 0, 1, 1)
+
+#if 0
+# define sntppkt_dbg(...) printf(__VA_ARGS__)
+#else
+# define sntppkt_dbg(...)
+#endif
+
+typedef struct
+{
+  uint32_t sec;
+  uint32_t frac;
+} ntp_timestamp_t;
+
+static const uint32_t NTP_TO_UNIX_EPOCH = 2208988800ul;
+
+typedef struct
+{
+  uint8_t mode : 3;
+  uint8_t ver : 3;
+  uint8_t LI : 2;
+  uint8_t stratum;
+  uint8_t poll;
+  uint8_t precision;
+  uint32_t delta_r;
+  uint32_t epsilon_r;
+  uint32_t refid;
+  ntp_timestamp_t ref;
+  ntp_timestamp_t origin;
+  ntp_timestamp_t recv;
+  ntp_timestamp_t xmit;
+} __attribute__((packed)) ntp_frame_t;
+
+#define NTP_RESPONSE_METATABLE  "sntppkt.resp"
+typedef struct {
+    /* Copied from incoming packet */
+    uint32_t  delta_r;
+    uint32_t  epsilon_r;
+    uint8_t   LI;
+    uint8_t   stratum;
+
+    /* Computed as per RFC 5905; units are 2^(-32) seconds */
+    int64_t   theta;
+    int64_t   delta;
+
+    /* Local computation */
+    uint64_t  t_rx;
+    uint32_t  score;
+} ntp_response_t;
+
+static const uint32_t MICROSECONDS      = 1000000;
+
+static uint64_t
+sntppkt_div1m(uint64_t n) {
+  uint64_t q1 = (n >> 5) + (n >> 10);
+  uint64_t q2 = (n >> 12) + (q1 >> 1);
+  uint64_t q3 = (q2 >> 11) - (q2 >> 23);
+
+  uint64_t q = n + q1 + q2 - q3;
+
+  q = q >> 20;
+
+  // Ignore the error term -- it is measured in pico seconds
+  return q;
+}
+
+static uint32_t
+sntppkt_us_to_frac(uint64_t us) {
+  return sntppkt_div1m(us << 32);
+}
+
+static uint32_t
+sntppkt_long_s(uint64_t time) {
+  return time >> 32;
+}
+
+static uint32_t
+sntppkt_long_us(uint64_t time) {
+  return ((time & 0xFFFFFFFF) * MICROSECONDS) >> 32;
+}
+
+/*
+ * Convert sec/usec to a Lua string suitable for depositing into a SNTP packet
+ * buffer.  This is a little gross, but it's not the worst thing a C
+ * programmer's ever done, I'm sure.
+ */
+static int
+sntppkt_make_ts(lua_State *L) {
+  ntp_timestamp_t ts;
+  uint32_t usec;
+
+  ts.sec  = htonl(luaL_checkinteger(L, 1) + NTP_TO_UNIX_EPOCH) ;
+  usec    =       luaL_checkinteger(L, 2)  ;
+  ts.frac = htonl(sntppkt_us_to_frac(usec));
+
+  lua_pushlstring(L, (const char *)&ts, sizeof(ts));
+  return 1;
+}
+
+/*
+ * Convert ntp_timestamp_t to uint64_t
+ */
+static uint64_t
+sntppkt_ts2uint64(ntp_timestamp_t ts) {
+  return (((uint64_t)(ts.sec)) << 32) + (uint64_t)(ts.frac);
+}
+
+/*
+ * Process a SNTP packet as contained in a Lua string, given a cookie timestamp
+ * and local clock second*usecond pair.  Generates a ntp_response_t userdata
+ * for later processing or a string if the server is telling us to go away.
+ *
+ * :: string (packet)
+ * -> string (cookie)
+ * -> int (local clock, sec component)
+ * -> int (local clock, usec component)
+ * -> sntppkt.resp
+ *
+ */
+static int
+sntppkt_proc_pkt(lua_State *L) {
```

Comment:
The Lua module style tends to be a lot more terse.  For example:
```C
  size_t       pkts_len, cookie_len;

  const char  *pkts       = lua_tolstring(L, 1, &pkts_len);
  ntp_timestamp_t *cookie = (ntp_timestamp_t*) lua_tolstring(L, 2, &cookie_len);
  uint32_t     now_sec    = luaL_checkinteger(L, 3);
  uint32_t     now_usec   = luaL_checkinteger(L, 4)
  ntp_frame_t  pktb;
  
  luaL_argcheck(L, pkts && pkts_len == sizeof(*pkts)), 1, "Bad packet");
  memcpy(&pktb, pkts, sizeof(*pkts));
  
  luaL_argcheck(L, cookie && pkts_len == sizeof(*cookie))
                   !memcmp(cookie, &pktb.origin, sizeof (*cookie)),
                 , 1, "Bad cookie");
  
  ntp_response_t *ntpr = lua_newuserdata(L, sizeof(ntp_response_t));;
...
```
Also note:
-  Try to use `luaL_argcheck()` to check args.
-  You can hoist the userdata create to the declaration preamble.  This works because its being returned to the caller as a result is what prevents is GC.  If an error is thrown or it isn't returned then it is collected.
-  Use the standard library synonyms when possible, e.g. `memcpy()` instead of `os_memcpy()`
-  Pre SDK 3, the runtime included an eLua "optimisation" where string constants in Flash where stored as a `const char *` reference in the `TString`  instead of a byte sequence.  Hence something like `('12345'):sub(2)` would not be word aligned.  I removed this because in practice it rarely saved much memory and there was a per access cost of handling the variant.

Reply:
AIUI

> This works because its being returned to the caller as a result is what prevents is GC.

is not quite right.  What's preventing the UD handle from getting GC'd is the existence of a reference in the registry.

By contrast, were `sntppkt_proc_pkt` to call up into Lua after its call to `lua_newuserdata` (and lack of `luaL_ref`) and encounter an exception that `longjmp()`-ed back over `sntppkt_proc_pkt`, say, just before returning back to Lua... the reason the UD would be eventually GC'd, despite having been returned to C and had $DEITY-knows-what done to it, is that there are no more references from the Lua side.